### PR TITLE
[SPARK-19306][Core] Fix inconsistent state in DiskBlockObject when expection occurred

### PR DESCRIPTION
**Pull request by jerryshao**
## What changes were proposed in this pull request?

In `DiskBlockObjectWriter`, when some errors happened during writing, it will call `revertPartialWritesAndClose`, if this method again failed due to some issues like out of disk, it will throw exception without resetting the state of this writer, also skipping the revert. So here propose to fix this issue to offer user a chance to recover from such issue.

## How was this patch tested?

Existing test.

### CODE REVIEW

**Comment by jerryshao** [comment]
CC @mridulm please help to review, thanks a lot.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala [hunk 1]
```diff
@@ -206,18 +209,22 @@ private[spark] class DiskBlockObjectWriter(
...
+    try {
+      truncateStream = new FileOutputStream(file, true)
+      truncateStream.getChannel.truncate(committedPosition)
+      file
     } catch {
       case e: Exception =>
         logError("Uncaught exception while reverting partial writes to file " + file, e)
         file
+    } finally {
+      truncateStream.close()
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala]: Another option is to use `Utils.tryWithSafeFinally` and do the if block with closeResources in the block with truncate in the finally block

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala]: You actually can't call `close()` here without a null check.

I'd also move the `file` return out of the try/catch block altogether.

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala]: Sorry about it. I will fix it.

**Comment by mridulm** [comment]
LGTM, will wait for @vanzin's comments too.

**Review by mridulm — APPROVED** [review]

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by vanzin** [comment]
LGTM. Merging to master / 2.1.

**Pull request closed without merging**